### DISCUSSION TITLE
[Solved (differently)] Array operation differentiation in an efficient way

### DISCUSSION THREAD

**Discussion by Mikolaj** [Mikolaj/mostly-harmless#11] [Ideas]
Edit: For the time being, this discussion is moved to https://github.com/Mikolaj/horde-ad/issues/4 and the emerging document is at https://github.com/Mikolaj/horde-ad/wiki/Extending-Delta-to-primitive-Vector-type

--

We already compute gradients of array sums or multiplications point-wise, using Real scalars. However, our benchmarks indicate it's much faster to treat arrays as a standalone differentiable primitive type. 

Note that it would already help if we had selectively fused gradient expressions for arrays operations. However, such expressions would grow as the sizes of argument arrays grow, so this is not even asymptotically optimal. Also, this is very hard to do both efficiently and generally, as seen in #6. If we implement array primitives, we can settle on generality, efficiency provided by the arrays.

Section 8.3 ("Arrays") suggests to treat arrays "as new, primitive, differentiable data types alongside R". I think we should tackle this early. This is only an optimization, but an indispensable one for ML applications. Even more importantly, this is, at least in my mind, tangled with the generalization to arbitrary shapes of scalar containers brainstormed in #6. Once we can handle arbitrary arrays as the new primitive types, perhaps the only really necessary generalization is to be able to handle tuples of primitive types. And we already can do the latter via vectors or lists --- not safely (we'd need all arrays of arbitrary rank and size to be of the same type), but that's a start. Moreover, such tuples (vectors) would not be 10^8 any more (individual arrays would be), so they don't need to be unboxed; the only thing that's probably still necessary is to keep them mutable, for `eval`, but that doesn't seem a problem.

I guess we would need to extend the `Delta` type? Let's experiment by expressing MNIST with that, benchmark, come back more grounded.

**Comment by Mikolaj** [comment]
Perhaps extending `Delta` nor changing `eval` is not need after all, if one squints really hard. I'm just trying a variant that uses `Array r` from the `hTensor` package. suggested by @awf, in place of scalars. Real scalars can be represented by rank 0 `Array r`. Let's see how it goes. There may be some fundamental problems, but for now I'm stuck at `hTensor` offering even less array operations than `hmatrix`, so I need to define them (and gradients of some of them). [Edit, actually there's enough operations, though some require conversions. There are missing gradients, though, see below.]

**Comment by Mikolaj** [comment]
Failure. Types seem fine, but current Delta seems not enough to express all necessary gradients. Citing two commit messages no branch `hTensor`:

--

* Switch to inputs that are boxed vectors of 'Array r' from hTensor

The idea is that Doubles will be represented as zero-rank arrays,
but the bulk of computations will be done on 1 and 2 rank arrays.
'Delta' datatype doesn't need to be extended and 'eval' is unchanged.

This compiles (after hTensor is fixed a little), but the array operations
are wrong or undefined, because various array operations and their
gradients need to be defined first. This may still not work,
because I may be missing something fundamental. Let's see!

--

* Finish the failed engine and MNIST nn definition overhaul to Arrays

OK, I've gone as far as I could --- it compiles and if wired to test harness, would probably produce correct function values, after some debugging. However, some of the gradient definitions are bogus, e.g., for the operation that sums elements of a vector and the operation that produces a vector of n copies of element x. Both of these operations are required for the softMax activation function in MNIST neural network. Unfortunately, I think Delta has to be extended to handle those and probably others.

The hTensor.Array approach in this failed prototype indeed solves the typing problem, even though at the cost of a very error-prone API, because array of all ranks and sizes have the same type. I think this may be a good setup to handle the extension to Arrays, if we choose to make it, sidestepping for a while the issue of expressing sums and products of well-typed vectors, matrices and tensors.

--

The mentioned inexpressible gradients (names taken from hmatrix, with ticks, as in backprop):

https://github.com/Mikolaj/mostly-harmless/blob/7ab28368e1392d8b1a28e78dc00c048e9ca6f1a2/src/ArrayAD.hs#L594-L604

**Comment by Mikolaj** [comment]
I guess `sumElements` on the `Delta` level can be had from `Dot`, the dot product operation, appearing as a grayed out `Delta` type constructor in Simon's slides. And `konst` is recoverable trivially by scaling a variable (`Var` constructor of `Delta`) that always holds the `1` vector and is never included in the training (we ignore the derivative computed with respect this this variable).

However, while I could write the `Dot` case of the `eval` function in the direct style (generalized to arbitrary rank arrays, including rank 0; or perhaps I'd curry `Dot` and not need to handle scalar results), I'm stumped on how to write it in the reverse let style. I mean, how to do the transformation, the eureka step, visible in this part of this commit: https://github.com/Mikolaj/mostly-harmless/commit/c6f0624efbac5dc094ee291a822ecf52f5dd129c#diff-0e761bf36ffd3568c496f61b270f409822e28d9cc402412156f2799a469e025bL55-R68

Or is `SumElements` easier than `Dot`? What is the algebraic structure we are after? Will we also need a constructor for a transpose of a linear map and various kinds of products due not the need to explicitly stratify Delta depending on the dimension of the space in question? I'd hate to lose the simplicity and speed that the simple Delta gives us versus keeping large arbitrary ad-hoc things on the tape. OTOH, I have no clue what people keep on the tape and I'd prefer not to learn. :)

**Comment by Mikolaj** [comment]
Unfortunately, the product from the `Num` instance of `hTensor.Array`

https://github.com/albertoruiz/hTensor/blob/master/lib/Numeric/LinearAlgebra/Array.hs#L48

seems to be the Kronecker product

https://github.com/albertoruiz/hTensor/blob/master/lib/Numeric/LinearAlgebra/Array/Internal.hs#L187-L209

and not one of the products we really need, e.g., the one that corresponds to an application of the linear function represented by a matrix to a vector

https://github.com/haskell-numerics/hmatrix/blob/master/packages/base/src/Internal/Container.hs#L78-L99

That means that the `Num` instance doesn't suffice for tensor multiplication and so our dual number definitions of multiplication don't suffice and so probably `Delta` doesn't suffice. E.g., the dual number definition of the application matrix-vector product would probably be something like (expressed in `hTensor` and `hmatrix`, following the definition in library `backprop`)

```
(#>!) :: DualDelta (Array r) -> DualDelta (Array r) -> DualDelta (Array r)
(#>!) (D u u') (D v v') =
  let uAs = asMatrix u
      vAs = asVector v
  in D (uAs #> vAs) (Add (asMatrix u' `outer` vAs) (tr uAs #> asVector v'))
```

and, as suggested above, neither `outer` product nor the application `#>` look like they could be expressed by scaling and addition of tensors, because they neither are the Num multiplication nor a derivative of it.

So, we miss not only the dot product (or sumElements), but also other crucial operations.

**Comment by Mikolaj** [comment]
Which is also why library `ad` AFAICS can't do that, even using its slow most general API, the one not specialized to `Double`.
